lib/elixir/unicode/tokenizer.ex

```diff
@@ -112,137 +106,362 @@ defmodule String.Tokenizer do
+    for {codepoint, scriptsets} <- all_codepoints_to_scriptset, into: %{} do
+      {codepoint,
+       scriptsets
+       |> Enum.map(&Map.fetch!(scriptset_masks, &1))
+       |> Enum.reduce(bottom, &ScriptSet.union/2)}
     end
+
+  ##
+  ## Define functions and module attributes to access characters and their scriptsets
+  ##
+
+  @bottom bottom
+  @latin 1
+  @top top
+  @indexed_scriptsets sorted_scriptsets |> Enum.with_index(&{&2, &1}) |> Map.new()
+
+  latin = Map.fetch!(scriptset_masks, "Latin")
+
+  @highly_restrictive [
+    ScriptSet.union(latin, Map.fetch!(scriptset_masks, "Japanese")),
+    ScriptSet.union(latin, Map.fetch!(scriptset_masks, "Han with Bopomofo")),
+    ScriptSet.union(latin, Map.fetch!(scriptset_masks, "Korean"))
+  ]
+
+  # ScriptSet helpers. Inline instead of dispatching to ScriptSet for performance
+
+  @compile {:inline, ss_latin: 1, ss_intersect: 2}
+  defp ss_latin(ss), do: :erlang.band(ss, @latin)
+  defp ss_intersect(left, right), do: :erlang.band(left, right)
+
+  # Ascii helpers
+
+  @compile {:inline, ascii_upper?: 1, ascii_lower?: 1, ascii_continue?: 1}
+  defp ascii_upper?(entry), do: entry >= ?A and entry <= ?Z
+  defp ascii_lower?(entry), do: entry >= ?a and entry <= ?z
+  defp ascii_continue?(entry), do: entry >= ?0 and entry <= ?9
+
+  # Pattern is used as a performance check to end sooner before traversing unicode
+  for pattern <- ' \t\n\r!"#$%&\'()*+,-./:;<=>?@[]^`{|}~' do
+    defp ascii_pattern?(unquote(pattern)), do: true
   end
 
-  defp unicode_upper?(_), do: false
+  defp ascii_pattern?(_), do: false
+
+  # Unicode helpers
+  # We use ranges whenever possible to reduce bytecode size.
+
+  unicode_upper = Enum.map(unicode_upper, &{&1, Map.get(codepoints_to_mask, &1, top)})
+  unicode_start = Enum.map(unicode_start, &{&1, Map.get(codepoints_to_mask, &1, top)})
+  unicode_continue = Enum.map(unicode_continue, &{&1, Map.get(codepoints_to_mask, &1, top)})
+
+  rangify = fn [{head, scriptset} | tail] ->
+    {first, last, scriptset, acc} =
+      Enum.reduce(tail, {head, head, scriptset, []}, fn
+        {number, scriptset}, {first, last, scriptset, acc} when number == first - 1 ->
+          {number, last, scriptset, acc}
+
+        {number, scriptset}, {first, last, range_scriptset, acc} ->
+          {number, number, scriptset, [{first, last, range_scriptset} | acc]}
+      end)
 
-  range = rangify.(unicode_start)
+    [{first, last, scriptset} | acc]
+  end
 
-  for {first, last} <- range do
+  for {first, last, scriptset} <- rangify.(unicode_upper) do
     if first == last do
-      defp unicode_start?(unquote(first)), do: true
+      defp unicode_upper(unquote(first)), do: unquote(scriptset)
     else
-      defp unicode_start?(entry) when entry in unquote(first)..unquote(last), do: true
+      defp unicode_upper(entry) when entry in unquote(first)..unquote(last),
+        do: unquote(scriptset)
     end
   end
 
-  defp unicode_start?(_), do: false
+  defp unicode_upper(_), do: 0
```

Comment:
Same question a bunch of places:

There's a mix of functions that use `0` directly, and those that use `@bottom` -- if we're going to get `@bottom` from  `lattice(size)` then should we use that everywhere? I don't know if module attrs have a runtime hit though or if they're compiled in.

Reply:
Looks like they're compiled in, from the terminology used in `h Kernel.@` ('takes a snapshot' among other phrases), so should be safe to change those all to `@bottom`; I tested doing that locally and tests still passed, but I can't push to this branch. Very low-importance change obviously.

Reply:
Oh, I was considered them to be unrelated “zeros” but treating it as a `@bottom` makes semantic sense too, as no codepoint should have no script set.